docs/Contributing/Blueprints/Troubleshooting-Hosts.md

```diff
@@ -0,0 +1,195 @@
+3. Osquery host executes the queries sequentially.
+4. Osquery fails to send the `distributed/write` request, e.g. because:
+    - proxies rejecting requests due to body size (which could happen if a lot of results are returned).
+    - temporary network errors (`distributed/read` succeeded, but `distributed/write` failed because of network partition).
+
+One way to detect these two scenarios (A) and (B) would be to check the following when executing the "distributed/read" request:
+
+```go
+// detectHostNotResponding returns whether the host hasn't been submitting results for sent queries.
+func (svc *Service) detectHostNotResponding(host *fleet.Host) bool {
+	interval := svc.config.Osquery.DetailUpdateInterval
+	if host.DistributedInterval > interval {
+		interval = host.DistributedInterval
+	}
+	// The following means Fleet hasn't received the distributed/request from the host
+	// "in a while", thus we assume the host is having some issue in executing the
+	// queries or sending the results.
+	return svc.clock.Now() > (host.DetailUpdatedAt + 2 * interval)
+} 
+```
+
+Notes: 
+- `Service.detectHostNotResponding` will allow us to detect such hosts (TBD: Where to store and show such info).
+- We use `2 * interval`, because of the artificial jitter added to the intervals in Fleet.
+- Default value for:
+    - host.DistributedInterval is usually 10s.
+    - svc.config.Osquery.DetailUpdateInterval is usually 1h.
+
+## Automatic Troubleshooting
+
+Once Fleet detects that a host is not responding the `distributed/read` with its corresponding
+`distributed/write`, we need a way to reduce the impact of the sent queries on such host.
+For instance, if a single query is the problematic one, then it doesn't make sense for all the other queries to miss execution on the host.
+
+### Host Vitals and Detail Queries
+
+What we call "Host Vitals" in Fleet could be split into two groups of queries:
+- Queries that update the `hosts` table (aka detail queries): 
+    - `network_interface`
+    - `os_version`
+    - `osquery_flags`
+    - `osquery_info`
+    - `system_info`
+    - `uptime`
+    - `disk_space_unix|disk_space_windows`
+- Queries that update tables that link to `hosts`:
+    - `mdm`
+    - `munki_info`
+    - `google_chrome_profiles`
+    - `orbit_info`
+    - `software_macos|software_linux|software_windows`
+    - `users`
+    - `scheduled_query_stats`
+	
+### Algorithm
+
+- The algorithm will assume the detail queries (first group) will never have any performance issues.
+Therefore we use them as the proof that `distributed/write` succeeded (by reading and updating
+`host.details_updated_at`).
+- The algorithm runs in the `distributed/read` and `distributed/write` requests (and we cannot assume that the server that processes the read request will be the same that processes the write request).
+- To not generate more load we will have a new column in `hosts` called something like `trouble_queries`.
```

Comment:
This will be a fun migration

Reply:
Alternative is using Redis to store, but incurring Redis overhead on every `distributed/read`.

Reply:
Ah good point.